src/install/dracut-install.c

```diff
@@ -559,10 +559,55 @@
         log_debug("%s %s", ldd, fullsrcpath);
         pid_t ldd_pid;
         if ((ldd_pid = fork()) == 0) {
+                char **cmdline = NULL;
+                int start, pos, idx = 0;
+                /* Account for at least 2 elements plus the terminating NULL in cmdline */
+                int args = 3;
+
+                /* Estimate the number of space-separated elements in the "ldd" string */
+                pos = 0;
+                while (isspace(ldd[pos]))
+                        pos++;
+                for (; ldd[pos]; pos++) {
+                        if (isspace(ldd[pos])) {
+                                if (pos)
+                                        args++;
+                                while (isspace(ldd[pos]))
+                                        pos++;
+                        }
+                }
+
+                cmdline = malloc(args * sizeof(char *));
+                memset(cmdline, 0, args * sizeof(char *));
+
+                pos = 0;
+                while (isspace(ldd[pos]))
+                        pos++;
+                start = pos;
+                for (; ldd[pos]; pos++) {
+                        while (ldd[pos] && !isspace(ldd[pos]))
+                                pos++;
+
+                        cmdline[idx] = malloc(pos - start + 1);
+                        memcpy(cmdline[idx], ldd + start, pos - start);
+                        cmdline[idx][pos - start] = 0;
+                        idx++;
+
+                        if (!ldd[pos])
+                                break;
+
+                        while (isspace(ldd[pos]))
+                                pos++;
+                        start = pos;
+                }
+
+                cmdline[idx++] = fullsrcpath;
+                cmdline[idx] = NULL;
+
                 dup2(fds[1], 1);
                 dup2(fds[1], 2);
                 putenv("LC_ALL=C");
-                execlp(ldd, ldd, fullsrcpath, (char *)NULL);
+                execvp(cmdline[0], cmdline);
```

Comment:
As I said, that's the whole point. It's not a security hole, but a feature to allow using a cross-compiled sysroot and a different ldd script that works on alien binaries and libraries.

Reply:
> It's not a security hole, but a feature

Perhaps both. dracut tents to be executed as root, so we should be considering mitigations against this security concern. Is there a way to validate the input (e.g. DRACUT_LDD) or perhaps the environment (e.g. --sysroot) ? Perhaps at least issue a warning in suspicious case ?

Reply:
> > It's not a security hole, but a feature
> 
> Perhaps both. dracut tents to be executed as root, so we should be considering mitigations against this security concern. Is there a way to validate the input (e.g. DRACUT_LDD) or perhaps the environment (e.g. --sysroot) ? Perhaps at least issue a warning in suspicious case ?

How do you propose to detect a suspicious case? By simply detecting either `--root` or `--sysroot` in case the command has arguments? How about a fake ldd that uses these arguments that's actually some kind of rootkit?

I think it's on the users to prevent this, i.e. the rpm scriptlet for the Red Hat / Fedora kernels should just use
```
unset DRACUT_LDD
```
before executing `dracut` because they don't use this feature but `dnf` still passes the whole environment to package scriptlets so the danger may be valid.

The cross-compiled sysroot feature existed before the dracut-ng fork, and I would like to continue using it.

¯\_(ツ)_/¯

Reply:
Would it be possible to ```unset DRACUT_LDD``` https://github.com/dracut-ng/dracut-ng/blob/main/install.d/50-dracut.install#L73 ? 

Would it be possible to document the security implications of setting ```DRACUT_LDD``` https://github.com/dracut-ng/dracut-ng/blob/main/man/dracut.8.asc?plain=1#L627

Reply:
> Would it be possible to `unset DRACUT_LDD` https://github.com/dracut-ng/dracut-ng/blob/main/install.d/50-dracut.install#L73 ?

Only if executing `dracut` manually from the command line or a script doesn't involve sourcing that file.

> Would it be possible to document the security implications of setting `DRACUT_LDD` https://github.com/dracut-ng/dracut-ng/blob/main/man/dracut.8.asc?plain=1#L627

That's certainly possible.

Reply:
Missed `strv_free(cmdline);` after the `execvp()`, in case it fails (for completeness).

Reply:
This one works fine here (using `_cleanup_strv_free_`):

```diff
diff --git a/src/install/dracut-install.c b/src/install/dracut-install.c
index 485143a5..58eff698 100644
--- a/src/install/dracut-install.c
+++ b/src/install/dracut-install.c
@@ -559,10 +559,16 @@ static int resolve_deps(const char *src)
         log_debug("%s %s", ldd, fullsrcpath);
         pid_t ldd_pid;
         if ((ldd_pid = fork()) == 0) {
+                _cleanup_strv_free_ char **cmdline = strv_split(ldd, WHITESPACE);
+
+                if (cmdline == NULL || strv_extend(&cmdline, fullsrcpath) < 0)
+                        /* consider as a hard error, see waitpid() below */
+                        _exit(126);
+
                 dup2(fds[1], 1);
                 dup2(fds[1], 2);
                 putenv("LC_ALL=C");
-                execlp(ldd, ldd, fullsrcpath, (char *)NULL);
+                execvp(cmdline[0], cmdline);
                 _exit(errno == ENOENT ? 127 : 126);
         }
         close(fds[1]);
```